`libgccjit.so` internal error when using `antoyo/gcc` and most recent commit

## Problem

With a freshly-compiled `libgccjit.so` and following the instructions in the [Building](https://github.com/antoyo/rustc_codegen_gcc#building) section of the README.md, my `rustc` gets stuck at building the `sysroot` with the following error (invoked with `cargo build -v`):

```sh
william@xubuntu-dtrain:~/Projects/toolchains/rustc_codegen_gcc$ ./test.sh --release
    Finished release [optimized] target(s) in 0.03s
[BUILD] mini_core
[BUILD] example
[AOT] mini_core_hello_world
./target/out/mini_core_hello_world
abc
bcd
Hello printf
Hello
World!
abc
Boxed outer got dropped!
Inner got dropped!
Inner got dropped!
Inner got dropped!
TLS works!

[BUILD] sysroot
    Updating crates.io index
   Compiling core v0.0.0 (/home/william/Projects/toolchains/rustc_codegen_gcc/build_sysroot/sysroot_src/library/core)
       Fresh cc v1.0.68
       Fresh autocfg v1.0.1
   Compiling compiler_builtins v0.1.45
     Running `rustc --crate-name core --edition=2018 sysroot_src/library/core/src/lib.rs --error-format=json --json=diagnostic-rendered-ansi,artifacts --crate-type lib --emit=dep-info,metadata,link -C embed-bitcode=no -C debuginfo=2 -C metadata=e063c7bddda6fe07 -C extra-filename=-e063c7bddda6fe07 --out-dir /home/william/Projects/toolchains/rustc_codegen_gcc/build_sysroot/target/x86_64-unknown-linux-gnu/debug/deps --target x86_64-unknown-linux-gnu -L dependency=/home/william/Projects/toolchains/rustc_codegen_gcc/build_sysroot/target/x86_64-unknown-linux-gnu/debug/deps -L dependency=/home/william/Projects/toolchains/rustc_codegen_gcc/build_sysroot/target/debug/deps -Cpanic=abort -Cdebuginfo=2 -Zpanic-abort-tests -Zcodegen-backend=/home/william/Projects/toolchains/rustc_codegen_gcc/target/release/librustc_codegen_gcc.so --sysroot /home/william/Projects/toolchains/rustc_codegen_gcc/build_sysroot/sysroot -Z force-unstable-if-unmarked -Cpanic=abort`
   Compiling libc v0.2.97
     Running `/home/william/Projects/toolchains/rustc_codegen_gcc/build_sysroot/target/debug/build/compiler_builtins-2759047e8e5760ee/build-script-build`
     Running `/home/william/Projects/toolchains/rustc_codegen_gcc/build_sysroot/target/debug/build/libc-7e3dc487b4357978/build-script-build`
   Compiling unwind v0.0.0 (/home/william/Projects/toolchains/rustc_codegen_gcc/build_sysroot/sysroot_src/library/unwind)
   Compiling miniz_oxide v0.4.4
   Compiling std v0.0.0 (/home/william/Projects/toolchains/rustc_codegen_gcc/build_sysroot/sysroot_src/library/std)
     Running `/home/william/Projects/toolchains/rustc_codegen_gcc/build_sysroot/target/debug/build/unwind-610dfb8f39c4878a/build-script-build`
     Running `/home/william/Projects/toolchains/rustc_codegen_gcc/build_sysroot/target/debug/build/miniz_oxide-713c8c4c6d9b1cdb/build-script-build`
     Running `/home/william/Projects/toolchains/rustc_codegen_gcc/build_sysroot/target/debug/build/std-1e5d80d9271fccf3/build-script-build`
   Compiling rustc-std-workspace-core v1.99.0 (/home/william/Projects/toolchains/rustc_codegen_gcc/build_sysroot/sysroot_src/library/rustc-std-workspace-core)
     Running `rustc --crate-name rustc_std_workspace_core --edition=2018 sysroot_src/library/rustc-std-workspace-core/lib.rs --error-format=json --json=diagnostic-rendered-ansi,artifacts --crate-type lib --emit=dep-info,metadata,link -C embed-bitcode=no -C debuginfo=2 -C metadata=59ed975568390d45 -C extra-filename=-59ed975568390d45 --out-dir /home/william/Projects/toolchains/rustc_codegen_gcc/build_sysroot/target/x86_64-unknown-linux-gnu/debug/deps --target x86_64-unknown-linux-gnu -L dependency=/home/william/Projects/toolchains/rustc_codegen_gcc/build_sysroot/target/x86_64-unknown-linux-gnu/debug/deps -L dependency=/home/william/Projects/toolchains/rustc_codegen_gcc/build_sysroot/target/debug/deps --extern core=/home/william/Projects/toolchains/rustc_codegen_gcc/build_sysroot/target/x86_64-unknown-linux-gnu/debug/deps/libcore-e063c7bddda6fe07.rmeta -Cpanic=abort -Cdebuginfo=2 -Zpanic-abort-tests -Zcodegen-backend=/home/william/Projects/toolchains/rustc_codegen_gcc/target/release/librustc_codegen_gcc.so --sysroot /home/william/Projects/toolchains/rustc_codegen_gcc/build_sysroot/sysroot -Z force-unstable-if-unmarked -Cpanic=abort`
     Running `rustc --crate-name compiler_builtins /home/william/.cargo/registry/src/github.com-1ecc6299db9ec823/compiler_builtins-0.1.45/src/lib.rs --error-format=json --json=diagnostic-rendered-ansi,artifacts --crate-type lib --emit=dep-info,metadata,link -C embed-bitcode=no -C debuginfo=2 --cfg 'feature="compiler-builtins"' --cfg 'feature="core"' --cfg 'feature="default"' --cfg 'feature="rustc-dep-of-std"' -C metadata=17b0aa9941b1c5cf -C extra-filename=-17b0aa9941b1c5cf --out-dir /home/william/Projects/toolchains/rustc_codegen_gcc/build_sysroot/target/x86_64-unknown-linux-gnu/debug/deps --target x86_64-unknown-linux-gnu -L dependency=/home/william/Projects/toolchains/rustc_codegen_gcc/build_sysroot/target/x86_64-unknown-linux-gnu/debug/deps -L dependency=/home/william/Projects/toolchains/rustc_codegen_gcc/build_sysroot/target/debug/deps --extern core=/home/william/Projects/toolchains/rustc_codegen_gcc/build_sysroot/target/x86_64-unknown-linux-gnu/debug/deps/librustc_std_workspace_core-59ed975568390d45.rmeta --cap-lints allow -Cpanic=abort -Cdebuginfo=2 -Zpanic-abort-tests -Zcodegen-backend=/home/william/Projects/toolchains/rustc_codegen_gcc/target/release/librustc_codegen_gcc.so --sysroot /home/william/Projects/toolchains/rustc_codegen_gcc/build_sysroot/sysroot -Z force-unstable-if-unmarked -Cpanic=abort --cfg 'feature="unstable"'`
libgccjit.so: error: type mismatch in binary expression
unsigned long

signed int

signed int

_21 = _16 & _20;
libgccjit.so: error: type mismatch in binary expression
unsigned long

signed int

unsigned long

_22 = _13 + _21;
libgccjit.so: error: ‘verify_gimple’ failed
0x7f9a3eadceb4 verify_gimple_in_seq(gimple*)
        ../../../gcc/gcc/tree-cfg.c:5161
0x7f9a3e7c2aa5 gimplify_body(tree_node*, bool)
        ../../../gcc/gcc/gimplify.c:15383
0x7f9a3e7c2c1c gimplify_function_tree(tree_node*)
        ../../../gcc/gcc/gimplify.c:15454
0x7f9a3e541e8e gcc::jit::playback::function::postprocess()
        ../../../gcc/gcc/jit/jit-playback.c:1727
0x7f9a3e542b5e gcc::jit::playback::context::replay()
        ../../../gcc/gcc/jit/jit-playback.c:3063
Please submit a full bug report,
with preprocessed source if appropriate.
Please include the complete backtrace with any bug report.
See <https://gcc.gnu.org/bugs/> for instructions.
error: could not compile `compiler_builtins`

Caused by:
  process didn't exit successfully: `rustc --crate-name compiler_builtins /home/william/.cargo/registry/src/github.com-1ecc6299db9ec823/compiler_builtins-0.1.45/src/lib.rs --error-format=json --json=diagnostic-rendered-ansi,artifacts --crate-type lib --emit=dep-info,metadata,link -C embed-bitcode=no -C debuginfo=2 --cfg 'feature="compiler-builtins"' --cfg 'feature="core"' --cfg 'feature="default"' --cfg 'feature="rustc-dep-of-std"' -C metadata=17b0aa9941b1c5cf -C extra-filename=-17b0aa9941b1c5cf --out-dir /home/william/Projects/toolchains/rustc_codegen_gcc/build_sysroot/target/x86_64-unknown-linux-gnu/debug/deps --target x86_64-unknown-linux-gnu -L dependency=/home/william/Projects/toolchains/rustc_codegen_gcc/build_sysroot/target/x86_64-unknown-linux-gnu/debug/deps -L dependency=/home/william/Projects/toolchains/rustc_codegen_gcc/build_sysroot/target/debug/deps --extern core=/home/william/Projects/toolchains/rustc_codegen_gcc/build_sysroot/target/x86_64-unknown-linux-gnu/debug/deps/librustc_std_workspace_core-59ed975568390d45.rmeta --cap-lints allow -Cpanic=abort -Cdebuginfo=2 -Zpanic-abort-tests -Zcodegen-backend=/home/william/Projects/toolchains/rustc_codegen_gcc/target/release/librustc_codegen_gcc.so --sysroot /home/william/Projects/toolchains/rustc_codegen_gcc/build_sysroot/sysroot -Z force-unstable-if-unmarked -Cpanic=abort --cfg 'feature="unstable"'` (exit status: 4)
warning: build failed, waiting for other jobs to finish...
libgccjit.so: error: type mismatch in binary expression
unsigned long

signed int

signed int

_34 = _28 & _33;
libgccjit.so: error: type mismatch in binary expression
unsigned long

signed int

unsigned long

_35 = _24 + _34;
libgccjit.so: error: ‘verify_gimple’ failed
0x7f646f068eb4 verify_gimple_in_seq(gimple*)
        ../../../gcc/gcc/tree-cfg.c:5161
0x7f646ed4eaa5 gimplify_body(tree_node*, bool)
        ../../../gcc/gcc/gimplify.c:15383
0x7f646ed4ec1c gimplify_function_tree(tree_node*)
        ../../../gcc/gcc/gimplify.c:15454
0x7f646eacde8e gcc::jit::playback::function::postprocess()
        ../../../gcc/gcc/jit/jit-playback.c:1727
0x7f646eaceb5e gcc::jit::playback::context::replay()
        ../../../gcc/gcc/jit/jit-playback.c:3063
Please submit a full bug report,
with preprocessed source if appropriate.
Please include the complete backtrace with any bug report.
See <https://gcc.gnu.org/bugs/> for instructions.
error: build failed

real    0m11.133s
user    0m12.593s
sys     0m0.514s
```

## Debugger Info

Unfortunately, running `gdb` doesn't give me any useful information; `loc` is a `NULL` pointer by the time we reach `add_error`:

```sh
william@xubuntu-dtrain:~/Projects/toolchains/rustc_codegen_gcc$ LD_LIBRARY_PATH=$(cat gcc_path) gdb --args rustc --crate-name compiler_builtins /home/william/.cargo/registry/src/github.com-1ecc6299db9ec823/compiler_builtins-0.1.45/src/lib.rs --error-format=json --json=diagnostic-rendered-ansi,artifacts --crate-type lib --emit=dep-info,metadata,link -C embed-bitcode=no -C debuginfo=2 --cfg 'feature="compiler-builtins"' --cfg 'feature="core"' --cfg 'feature="default"' --cfg 'feature="rustc-dep-of-std"' -C metadata=17b0aa9941b1c5cf -C extra-filename=-17b0aa9941b1c5cf --out-dir /home/william/Projects/toolchains/rustc_codegen_gcc/build_sysroot/target/x86_64-unknown-linux-gnu/debug/deps --target x86_64-unknown-linux-gnu -L dependency=/home/william/Projects/toolchains/rustc_codegen_gcc/build_sysroot/target/x86_64-unknown-linux-gnu/debug/deps -L dependency=/home/william/Projects/toolchains/rustc_codegen_gcc/build_sysroot/target/debug/deps --extern core=/home/william/Projects/toolchains/rustc_codegen_gcc/build_sysroot/target/x86_64-unknown-linux-gnu/debug/deps/librustc_std_workspace_core-59ed975568390d45.rmeta --cap-lints allow -Cpanic=abort -Cdebuginfo=2 -Zpanic-abort-tests -Zcodegen-backend=/home/william/Projects/toolchains/rustc_codegen_gcc/target/release/librustc_codegen_gcc.so --sysroot /home/william/Projects/toolchains/rustc_codegen_gcc/build_sysroot/sysroot -Z force-unstable-if-unmarked -Cpanic=abort --cfg 'feature="unstable"'
GNU gdb (Ubuntu 9.2-0ubuntu1~20.04) 9.2
Copyright (C) 2020 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "x86_64-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from rustc...
(gdb) b add_error
Function "add_error" not defined.
Make breakpoint pending on future shared library load? (y or [n]) y
Breakpoint 1 (add_error) pending.
(gdb) c
The program is not being run.
(gdb) r
Starting program: /home/william/.cargo/bin/rustc --crate-name compiler_builtins /home/william/.cargo/registry/src/github.com-1ecc6299db9ec823/compiler_builtins-0.1.45/src/lib.rs --error-format=json --json=diagnostic-rendered-ansi,artifacts --crate-type lib --emit=dep-info,metadata,link -C embed-bitcode=no -C debuginfo=2 --cfg feature=\"compiler-builtins\" --cfg feature=\"core\" --cfg feature=\"default\" --cfg feature=\"rustc-dep-of-std\" -C metadata=17b0aa9941b1c5cf -C extra-filename=-17b0aa9941b1c5cf --out-dir /home/william/Projects/toolchains/rustc_codegen_gcc/build_sysroot/target/x86_64-unknown-linux-gnu/debug/deps --target x86_64-unknown-linux-gnu -L dependency=/home/william/Projects/toolchains/rustc_codegen_gcc/build_sysroot/target/x86_64-unknown-linux-gnu/debug/deps -L dependency=/home/william/Projects/toolchains/rustc_codegen_gcc/build_sysroot/target/debug/deps --extern core=/home/william/Projects/toolchains/rustc_codegen_gcc/build_sysroot/target/x86_64-unknown-linux-gnu/debug/deps/librustc_std_workspace_core-59ed975568390d45.rmeta --cap-lints allow -Cpanic=abort -Cdebuginfo=2 -Zpanic-abort-tests -Zcodegen-backend=/home/william/Projects/toolchains/rustc_codegen_gcc/target/release/librustc_codegen_gcc.so --sysroot /home/william/Projects/toolchains/rustc_codegen_gcc/build_sysroot/sysroot -Z force-unstable-if-unmarked -Cpanic=abort --cfg feature=\"unstable\"
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
process 517958 is executing new program: /home/william/.rustup/toolchains/nightly-2021-06-14-x86_64-unknown-linux-gnu/bin/rustc
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
[New Thread 0x7fffee3ff700 (LWP 517962)]
{"artifact":"/home/william/Projects/toolchains/rustc_codegen_gcc/build_sysroot/target/x86_64-unknown-linux-gnu/debug/deps/compiler_builtins-17b0aa9941b1c5cf.d","emit":"dep-info"}
{"artifact":"/home/william/Projects/toolchains/rustc_codegen_gcc/build_sysroot/target/x86_64-unknown-linux-gnu/debug/deps/libcompiler_builtins-17b0aa9941b1c5cf.rmeta","emit":"metadata"}
[New Thread 0x7fffe678a700 (LWP 517963)]
[New Thread 0x7fffe6589700 (LWP 517964)]
[New Thread 0x7fffe59ff700 (LWP 517965)]
[New Thread 0x7fffe4a8f700 (LWP 517966)]
[New Thread 0x7fffe3fff700 (LWP 517967)]
[New Thread 0x7fffe37ff700 (LWP 517968)]
[New Thread 0x7fffe25ff700 (LWP 517969)]
[New Thread 0x7fffe1dff700 (LWP 517970)]
[New Thread 0x7fffe15ff700 (LWP 517971)]
[New Thread 0x7fffe0b7f700 (LWP 517972)]
[New Thread 0x7fffe03ff700 (LWP 517973)]
[New Thread 0x7fffdfbff700 (LWP 517974)]
[New Thread 0x7fffdf3ff700 (LWP 517975)]
[New Thread 0x7fffdebff700 (LWP 517976)]
[New Thread 0x7fffde1ff700 (LWP 517977)]
[New Thread 0x7fffdd9ff700 (LWP 517978)]
[New Thread 0x7fffdd1ff700 (LWP 517979)]
[New Thread 0x7fffdc9ff700 (LWP 517980)]
[Detaching after vfork from child process 517981]
[Thread 0x7fffe59ff700 (LWP 517965) exited]
[Detaching after vfork from child process 517982]
[Thread 0x7fffe4a8f700 (LWP 517966) exited]
--Type <RET> for more, q to quit, c to continue without paging--
[Switching to Thread 0x7fffe3fff700 (LWP 517967)]

Thread 7 "opt compiler_bu" hit Breakpoint 1, gcc::jit::recording::context::add_error (
    this=0x7fffe8886300, loc=0x0, fmt=fmt@entry=0x7fffecbf822d "%s")
    at ../../../gcc/gcc/jit/jit-recording.c:1457
1457    {
(gdb) p loc->m_line
Cannot access memory at address 0x28
(gdb) p loc
$1 = (gcc::jit::recording::location *) 0x0
(gdb) p loc->m_line
Cannot access memory at address 0x28
(gdb)
```

I am unsure how to proceed, so opening this issue to see whether I found an actual bug, or I'm doing something wrong.

## Additional Info

One thing I noticed... despite me passing `--release` to `test.sh`, the `cargo` command actually invoked in `build_sysroot.sh` is the debug invocation (I have added `-v`):

```sh
    sysroot_channel='debug'
    cargo build -v --target $TARGET_TRIPLE
```

## System Information

OS:
```sh
william@xubuntu-dtrain:~/Projects/toolchains/rustc_codegen_gcc$ uname -a
Linux xubuntu-dtrain 5.4.0-73-generic #82-Ubuntu SMP Wed Apr 14 17:39:42 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux
```

GCC version:
```sh
william@xubuntu-dtrain:~/Projects/toolchains/rustc_codegen_gcc$ gcc -v
Using built-in specs.
COLLECT_GCC=gcc
COLLECT_LTO_WRAPPER=/usr/lib/gcc/x86_64-linux-gnu/9/lto-wrapper
OFFLOAD_TARGET_NAMES=nvptx-none:hsa
OFFLOAD_TARGET_DEFAULT=1
Target: x86_64-linux-gnu
Configured with: ../src/configure -v --with-pkgversion='Ubuntu 9.3.0-17ubuntu1~20.04' --with-bugurl=file:///usr/share/doc/gcc-9/README.Bugs --enable-languages=c,ada,c++,go,brig,d,fortran,objc,obj-c++,gm2 --prefix=/usr --with-gcc-major-version-only --program-suffix=-9 --program-prefix=x86_64-linux-gnu- --enable-shared --enable-linker-build-id --libexecdir=/usr/lib --without-included-gettext --enable-threads=posix --libdir=/usr/lib --enable-nls --enable-clocale=gnu --enable-libstdcxx-debug --enable-libstdcxx-time=yes --with-default-libstdcxx-abi=new --enable-gnu-unique-object --disable-vtable-verify --enable-plugin --enable-default-pie --with-system-zlib --with-target-system-zlib=auto --enable-objc-gc=auto --enable-multiarch --disable-werror --with-arch-32=i686 --with-abi=m64 --with-multilib-list=m32,m64,mx32 --enable-multilib --with-tune=generic --enable-offload-targets=nvptx-none=/build/gcc-9-HskZEa/gcc-9-9.3.0/debian/tmp-nvptx/usr,hsa --without-cuda-driver --enable-checking=release --build=x86_64-linux-gnu --host=x86_64-linux-gnu --target=x86_64-linux-gnu
Thread model: posix
gcc version 9.3.0 (Ubuntu 9.3.0-17ubuntu1~20.04)
william@xubuntu-dtrain:~/Projects/toolchains/rustc_codegen_gcc$
```

`libgccjit` `configure` invocation:
```sh
../../gcc/configure --prefix=/usr --enable-languages=jit,c++ --disable-multilib --disable-bootstrap --enable-host-shared --enable-checking=release
```

`gcc-path` contents:
```sh
/home/william/Projects/toolchains/build-gcc-toolchain/destdir-libgccjit/usr/lib
```

`rustc` version (matches expected):
```sh
william@xubuntu-dtrain:~/Projects/toolchains/rustc_codegen_gcc$ rustc -V
rustc 1.55.0-nightly (f586d79d1 2021-06-13)
```

Here is the entire `gdb` dump; the file appears to be `compiler_builtins.1shhbuin-cgu.1.c`.

```sh
william@xubuntu-dtrain:~/Projects/toolchains/rustc_codegen_gcc$ LD_LIBRARY_PATH=$(cat gcc_path) gdb --args rustc --crate-name compiler_builtins /home/william/.cargo/registry/src/github.com-1ecc6299db9ec823/compiler_builtins-0.1.45/src/lib.rs --error-format=json --json=diagnostic-rendered-ansi,artifacts --crate-type lib --emit=dep-info,metadata,link -C embed-bitcode=no -C debuginfo=2 --cfg 'feature="compiler-builtins"' --cfg 'feature="core"' --cfg 'feature="default"' --cfg 'feature="rustc-dep-of-std"' -C metadata=17b0aa9941b1c5cf -C extra-filename=-17b0aa9941b1c5cf --out-dir /home/william/Projects/toolchains/rustc_codegen_gcc/build_sysroot/target/x86_64-unknown-linux-gnu/debug/deps --target x86_64-unknown-linux-gnu -L dependency=/home/william/Projects/toolchains/rustc_codegen_gcc/build_sysroot/target/x86_64-unknown-linux-gnu/debug/deps -L dependency=/home/william/Projects/toolchains/rustc_codegen_gcc/build_sysroot/target/debug/deps --extern core=/home/william/Projects/toolchains/rustc_codegen_gcc/build_sysroot/target/x86_64-unknown-linux-gnu/debug/deps/librustc_std_workspace_core-59ed975568390d45.rmeta --cap-lints allow -Cpanic=abort -Cdebuginfo=2 -Zpanic-abort-tests -Zcodegen-backend=/home/william/Projects/toolchains/rustc_codegen_gcc/target/release/librustc_codegen_gcc.so --sysroot /home/william/Projects/toolchains/rustc_codegen_gcc/build_sysroot/sysroot -Z force-unstable-if-unmarked -Cpanic=abort --cfg 'feature="unstable"'
GNU gdb (Ubuntu 9.2-0ubuntu1~20.04) 9.2
Copyright (C) 2020 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "x86_64-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from rustc...
(gdb) r
Starting program: /home/william/.cargo/bin/rustc --crate-name compiler_builtins /home/william/.cargo/registry/src/github.com-1ecc6299db9ec823/compiler_builtins-0.1.45/src/lib.rs --error-format=json --json=diagnostic-rendered-ansi,artifacts --crate-type lib --emit=dep-info,metadata,link -C embed-bitcode=no -C debuginfo=2 --cfg feature=\"compiler-builtins\" --cfg feature=\"core\" --cfg feature=\"default\" --cfg feature=\"rustc-dep-of-std\" -C metadata=17b0aa9941b1c5cf -C extra-filename=-17b0aa9941b1c5cf --out-dir /home/william/Projects/toolchains/rustc_codegen_gcc/build_sysroot/target/x86_64-unknown-linux-gnu/debug/deps --target x86_64-unknown-linux-gnu -L dependency=/home/william/Projects/toolchains/rustc_codegen_gcc/build_sysroot/target/x86_64-unknown-linux-gnu/debug/deps -L dependency=/home/william/Projects/toolchains/rustc_codegen_gcc/build_sysroot/target/debug/deps --extern core=/home/william/Projects/toolchains/rustc_codegen_gcc/build_sysroot/target/x86_64-unknown-linux-gnu/debug/deps/librustc_std_workspace_core-59ed975568390d45.rmeta --cap-lints allow -Cpanic=abort -Cdebuginfo=2 -Zpanic-abort-tests -Zcodegen-backend=/home/william/Projects/toolchains/rustc_codegen_gcc/target/release/librustc_codegen_gcc.so --sysroot /home/william/Projects/toolchains/rustc_codegen_gcc/build_sysroot/sysroot -Z force-unstable-if-unmarked -Cpanic=abort --cfg feature=\"unstable\"
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
process 519806 is executing new program: /home/william/.rustup/toolchains/nightly-2021-06-14-x86_64-unknown-linux-gnu/bin/rustc
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
[New Thread 0x7fffee3ff700 (LWP 519810)]
{"artifact":"/home/william/Projects/toolchains/rustc_codegen_gcc/build_sysroot/target/x86_64-unknown-linux-gnu/debug/deps/compiler_builtins-17b0aa9941b1c5cf.d","emit":"dep-info"}
{"artifact":"/home/william/Projects/toolchains/rustc_codegen_gcc/build_sysroot/target/x86_64-unknown-linux-gnu/debug/deps/libcompiler_builtins-17b0aa9941b1c5cf.rmeta","emit":"metadata"}
[New Thread 0x7fffe678a700 (LWP 519811)]
[New Thread 0x7fffe5fff700 (LWP 519812)]
[New Thread 0x7fffe57ff700 (LWP 519813)]
[New Thread 0x7fffe4fff700 (LWP 519814)]
[New Thread 0x7fffe3e0f700 (LWP 519815)]
[New Thread 0x7fffe37ff700 (LWP 519816)]
[New Thread 0x7fffe25ff700 (LWP 519817)]
[New Thread 0x7fffe1dff700 (LWP 519818)]
[New Thread 0x7fffe15ff700 (LWP 519819)]
[New Thread 0x7fffe0dff700 (LWP 519820)]
[New Thread 0x7fffe037f700 (LWP 519821)]
[New Thread 0x7fffdfbff700 (LWP 519822)]
[New Thread 0x7fffdf17f700 (LWP 519823)]
[New Thread 0x7fffde9ff700 (LWP 519824)]
[New Thread 0x7fffde1ff700 (LWP 519825)]
[New Thread 0x7fffdd9ff700 (LWP 519826)]
[New Thread 0x7fffdd1ff700 (LWP 519827)]
[New Thread 0x7fffdc9ff700 (LWP 519828)]
[Detaching after vfork from child process 519829]
[Thread 0x7fffe4fff700 (LWP 519814) exited]
[Detaching after vfork from child process 519830]
[Thread 0x7fffe37ff700 (LWP 519816) exited]
[Detaching after vfork from child process 519831]
[Thread 0x7fffe1dff700 (LWP 519818) exited]
[Detaching after vfork from child process 519832]
[Thread 0x7fffe25ff700 (LWP 519817) exited]
libgccjit.so: /tmp/dumps/compiler_builtins.1shhbuin-cgu.1.c:2219:0: error: type mismatch in binary expression
unsigned long

signed int

signed int

_21 = _16 & _20;
libgccjit.so: /tmp/dumps/compiler_builtins.1shhbuin-cgu.1.c:2219:0: error: type mismatch in binary expression
unsigned long

signed int

unsigned long

_22 = _13 + _21;
libgccjit.so: /tmp/dumps/compiler_builtins.1shhbuin-cgu.1.c:2219:0: error: ‘verify_gimple’ failed
0x7fffebe68eb4 verify_gimple_in_seq(gimple*)
        ../../../gcc/gcc/tree-cfg.c:5161
0x7fffebb4eaa5 gimplify_body(tree_node*, bool)
        ../../../gcc/gcc/gimplify.c:15383
0x7fffebb4ec1c gimplify_function_tree(tree_node*)
        ../../../gcc/gcc/gimplify.c:15454
0x7fffeb8cde8e gcc::jit::playback::function::postprocess()
        ../../../gcc/gcc/jit/jit-playback.c:1727
0x7fffeb8ceb5e gcc::jit::playback::context::replay()
        ../../../gcc/gcc/jit/jit-playback.c:3063
Please submit a full bug report,
with preprocessed source if appropriate.
Please include the complete backtrace with any bug report.
See <https://gcc.gnu.org/bugs/> for instructions.
[Thread 0x7fffdc9ff700 (LWP 519828) exited]
[Thread 0x7fffdd1ff700 (LWP 519827) exited]
[Thread 0x7fffdd9ff700 (LWP 519826) exited]
[Thread 0x7fffde1ff700 (LWP 519825) exited]
[Thread 0x7fffde9ff700 (LWP 519824) exited]
[Thread 0x7fffdf17f700 (LWP 519823) exited]
[Thread 0x7fffdfbff700 (LWP 519822) exited]
[Thread 0x7fffe037f700 (LWP 519821) exited]
[Thread 0x7fffe0dff700 (LWP 519820) exited]
[Thread 0x7fffe15ff700 (LWP 519819) exited]
[Thread 0x7fffe3e0f700 (LWP 519815) exited]
[Thread 0x7fffe57ff700 (LWP 519813) exited]
[Thread 0x7fffe5fff700 (LWP 519812) exited]
[Thread 0x7fffe678a700 (LWP 519811) exited]
[Thread 0x7fffeed40500 (LWP 519806) exited]
--Type <RET> for more, q to quit, c to continue without paging--
[Inferior 1 (process 519806) exited with code 04]
(gdb)
```

Yes :D!

```
william@xubuntu-dtrain:~/Projects/toolchains/rustc_codegen_gcc$ LD_LIBRARY_PATH=$(cat gcc_path) gdb --args rustc --crate-name compiler_builtins /home/william/.cargo/registry/src/github.com-1ecc6299db9ec823/compiler_builtins-0.1.45/src/lib.rs --error-format=json --json=diagnostic-rendered-ansi,artifacts --crate-type lib --emit=dep-info,metadata,link -C embed-bitcode=no -C debuginfo=2 --cfg 'feature="compiler-builtins"' --cfg 'feature="core"' --cfg 'feature="default"' --cfg 'feature="rustc-dep-of-std"' -C metadata=17b0aa9941b1c5cf -C extra-filename=-17b0aa9941b1c5cf --out-dir /home/william/Projects/toolchains/rustc_codegen_gcc/build_sysroot/target/x86_64-unknown-linux-gnu/debug/deps --target x86_64-unknown-linux-gnu -L dependency=/home/william/Projects/toolchains/rustc_codegen_gcc/build_sysroot/target/x86_64-unknown-linux-gnu/debug/deps -L dependency=/home/william/Projects/toolchains/rustc_codegen_gcc/build_sysroot/target/debug/deps --extern core=/home/william/Projects/toolchains/rustc_codegen_gcc/build_sysroot/target/x86_64-unknown-linux-gnu/debug/deps/librustc_std_workspace_core-59ed975568390d45.rmeta --cap-lints allow -Cpanic=abort -Cdebuginfo=2 -Zpanic-abort-tests -Zcodegen-backend=/home/william/Projects/toolchains/rustc_codegen_gcc/target/release/librustc_codegen_gcc.so --sysroot /home/william/Projects/toolchains/rustc_codegen_gcc/build_sysroot/sysroot -Z force-unstable-if-unmarked -Cpanic=abort --cfg 'feature="unstable"'
GNU gdb (Ubuntu 9.2-0ubuntu1~20.04) 9.2
Copyright (C) 2020 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "x86_64-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from rustc...
(gdb) r
Starting program: /home/william/.cargo/bin/rustc --crate-name compiler_builtins /home/william/.cargo/registry/src/github.com-1ecc6299db9ec823/compiler_builtins-0.1.45/src/lib.rs --error-format=json --json=diagnostic-rendered-ansi,artifacts --crate-type lib --emit=dep-info,metadata,link -C embed-bitcode=no -C debuginfo=2 --cfg feature=\"compiler-builtins\" --cfg feature=\"core\" --cfg feature=\"default\" --cfg feature=\"rustc-dep-of-std\" -C metadata=17b0aa9941b1c5cf -C extra-filename=-17b0aa9941b1c5cf --out-dir /home/william/Projects/toolchains/rustc_codegen_gcc/build_sysroot/target/x86_64-unknown-linux-gnu/debug/deps --target x86_64-unknown-linux-gnu -L dependency=/home/william/Projects/toolchains/rustc_codegen_gcc/build_sysroot/target/x86_64-unknown-linux-gnu/debug/deps -L dependency=/home/william/Projects/toolchains/rustc_codegen_gcc/build_sysroot/target/debug/deps --extern core=/home/william/Projects/toolchains/rustc_codegen_gcc/build_sysroot/target/x86_64-unknown-linux-gnu/debug/deps/librustc_std_workspace_core-59ed975568390d45.rmeta --cap-lints allow -Cpanic=abort -Cdebuginfo=2 -Zpanic-abort-tests -Zcodegen-backend=/home/william/Projects/toolchains/rustc_codegen_gcc/target/release/librustc_codegen_gcc.so --sysroot /home/william/Projects/toolchains/rustc_codegen_gcc/build_sysroot/sysroot -Z force-unstable-if-unmarked -Cpanic=abort --cfg feature=\"unstable\"
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
process 519981 is executing new program: /home/william/.rustup/toolchains/nightly-2021-06-14-x86_64-unknown-linux-gnu/bin/rustc
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
[New Thread 0x7fffee3ff700 (LWP 519985)]
{"artifact":"/home/william/Projects/toolchains/rustc_codegen_gcc/build_sysroot/target/x86_64-unknown-linux-gnu/debug/deps/compiler_builtins-17b0aa9941b1c5cf.d","emit":"dep-info"}
{"artifact":"/home/william/Projects/toolchains/rustc_codegen_gcc/build_sysroot/target/x86_64-unknown-linux-gnu/debug/deps/libcompiler_builtins-17b0aa9941b1c5cf.rmeta","emit":"metadata"}
[New Thread 0x7fffe678a700 (LWP 519986)]
[New Thread 0x7fffe6589700 (LWP 519987)]
[New Thread 0x7fffe59ff700 (LWP 519988)]
[New Thread 0x7fffe51ff700 (LWP 519989)]
[New Thread 0x7fffe400f700 (LWP 519990)]
[New Thread 0x7fffe39ff700 (LWP 519991)]
[New Thread 0x7fffe27ff700 (LWP 519992)]
[New Thread 0x7fffe1fff700 (LWP 519993)]
[New Thread 0x7fffe17ff700 (LWP 519994)]
[New Thread 0x7fffe0fff700 (LWP 519995)]
[New Thread 0x7fffe057f700 (LWP 519996)]
[New Thread 0x7fffdfb7f700 (LWP 519997)]
[New Thread 0x7fffdf3ff700 (LWP 519998)]
[New Thread 0x7fffdebff700 (LWP 519999)]
[New Thread 0x7fffde3ff700 (LWP 520000)]
[New Thread 0x7fffddbff700 (LWP 520001)]
[New Thread 0x7fffdd3ff700 (LWP 520002)]
[New Thread 0x7fffdc8ff700 (LWP 520003)]
[Detaching after vfork from child process 520004]
[Thread 0x7fffe51ff700 (LWP 519989) exited]
[Detaching after vfork from child process 520005]
[Thread 0x7fffe39ff700 (LWP 519991) exited]
[Detaching after vfork from child process 520006]
[Thread 0x7fffe1fff700 (LWP 519993) exited]
[Detaching after vfork from child process 520007]
[Thread 0x7fffe27ff700 (LWP 519992) exited]
[Detaching after vfork from child process 520008]
[Thread 0x7fffe17ff700 (LWP 519994) exited]
libgccjit.so: /tmp/dumps/compiler_builtins.1shhbuin-cgu.1.c:2219:0: error: type mismatch in binary expression
unsigned long

signed int

signed int

_21 = _16 & _20;
libgccjit.so: /tmp/dumps/compiler_builtins.1shhbuin-cgu.1.c:2219:0: error: type mismatch in binary expression
unsigned long

signed int

unsigned long

_22 = _13 + _21;
libgccjit.so: /tmp/dumps/compiler_builtins.1shhbuin-cgu.1.c:2219:0: error: ‘verify_gimple’ failed
0x7fffebe68eb4 verify_gimple_in_seq(gimple*)
        ../../../gcc/gcc/tree-cfg.c:5161
0x7fffebb4eaa5 gimplify_body(tree_node*, bool)
        ../../../gcc/gcc/gimplify.c:15383
0x7fffebb4ec1c gimplify_function_tree(tree_node*)
        ../../../gcc/gcc/gimplify.c:15454
0x7fffeb8cde8e gcc::jit::playback::function::postprocess()
        ../../../gcc/gcc/jit/jit-playback.c:1727
0x7fffeb8ceb5e gcc::jit::playback::context::replay()
        ../../../gcc/gcc/jit/jit-playback.c:3063
Please submit a full bug report,
with preprocessed source if appropriate.
Please include the complete backtrace with any bug report.
See <https://gcc.gnu.org/bugs/> for instructions.
[Thread 0x7fffdc8ff700 (LWP 520003) exited]
[Thread 0x7fffdd3ff700 (LWP 520002) exited]
[Thread 0x7fffddbff700 (LWP 520001) exited]
[Thread 0x7fffde3ff700 (LWP 520000) exited]
[Thread 0x7fffdebff700 (LWP 519999) exited]
[Thread 0x7fffdf3ff700 (LWP 519998) exited]
[Thread 0x7fffdfb7f700 (LWP 519997) exited]
[Thread 0x7fffe057f700 (LWP 519996) exited]
[Thread 0x7fffe0fff700 (LWP 519995) exited]
[Thread 0x7fffe400f700 (LWP 519990) exited]
[Thread 0x7fffe59ff700 (LWP 519988) exited]
[Thread 0x7fffe6589700 (LWP 519987) exited]
[Thread 0x7fffe678a700 (LWP 519986) exited]
[Thread 0x7fffeed40500 (LWP 519981) exited]
--Type <RET> for more, q to quit, c to continue without paging--
[Inferior 1 (process 519981) exited with code 04]
(gdb) b add_error
Breakpoint 1 at 0x7fffeb8bf240: add_error. (2 locations)
(gdb) r
Starting program: /home/william/.rustup/toolchains/nightly-2021-06-14-x86_64-unknown-linux-gnu/bin/rustc --crate-name compiler_builtins /home/william/.cargo/registry/src/github.com-1ecc6299db9ec823/compiler_builtins-0.1.45/src/lib.rs --error-format=json --json=diagnostic-rendered-ansi,artifacts --crate-type lib --emit=dep-info,metadata,link -C embed-bitcode=no -C debuginfo=2 --cfg feature=\"compiler-builtins\" --cfg feature=\"core\" --cfg feature=\"default\" --cfg feature=\"rustc-dep-of-std\" -C metadata=17b0aa9941b1c5cf -C extra-filename=-17b0aa9941b1c5cf --out-dir /home/william/Projects/toolchains/rustc_codegen_gcc/build_sysroot/target/x86_64-unknown-linux-gnu/debug/deps --target x86_64-unknown-linux-gnu -L dependency=/home/william/Projects/toolchains/rustc_codegen_gcc/build_sysroot/target/x86_64-unknown-linux-gnu/debug/deps -L dependency=/home/william/Projects/toolchains/rustc_codegen_gcc/build_sysroot/target/debug/deps --extern core=/home/william/Projects/toolchains/rustc_codegen_gcc/build_sysroot/target/x86_64-unknown-linux-gnu/debug/deps/librustc_std_workspace_core-59ed975568390d45.rmeta --cap-lints allow -Cpanic=abort -Cdebuginfo=2 -Zpanic-abort-tests -Zcodegen-backend=/home/william/Projects/toolchains/rustc_codegen_gcc/target/release/librustc_codegen_gcc.so --sysroot /home/william/Projects/toolchains/rustc_codegen_gcc/build_sysroot/sysroot -Z force-unstable-if-unmarked -Cpanic=abort --cfg feature=\"unstable\"

[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
[New Thread 0x7fffee3ff700 (LWP 520010)]
{"artifact":"/home/william/Projects/toolchains/rustc_codegen_gcc/build_sysroot/target/x86_64-unknown-linux-gnu/debug/deps/compiler_builtins-17b0aa9941b1c5cf.d","emit":"dep-info"}
{"artifact":"/home/william/Projects/toolchains/rustc_codegen_gcc/build_sysroot/target/x86_64-unknown-linux-gnu/debug/deps/libcompiler_builtins-17b0aa9941b1c5cf.rmeta","emit":"metadata"}
[New Thread 0x7fffe678a700 (LWP 520011)]
[New Thread 0x7fffe5fff700 (LWP 520012)]
[New Thread 0x7fffe57ff700 (LWP 520013)]
[New Thread 0x7fffe4fff700 (LWP 520014)]
[New Thread 0x7fffe3e0f700 (LWP 520015)]
[New Thread 0x7fffe37ff700 (LWP 520016)]
[New Thread 0x7fffe25ff700 (LWP 520017)]
[New Thread 0x7fffe1dff700 (LWP 520018)]
[New Thread 0x7fffe15ff700 (LWP 520019)]
[New Thread 0x7fffe0dff700 (LWP 520020)]
[New Thread 0x7fffe05ff700 (LWP 520021)]
[New Thread 0x7fffdfdff700 (LWP 520022)]
[New Thread 0x7fffdf37f700 (LWP 520023)]
[New Thread 0x7fffde97f700 (LWP 520024)]
[New Thread 0x7fffde1ff700 (LWP 520025)]
[New Thread 0x7fffdd9ff700 (LWP 520026)]
[New Thread 0x7fffdd1ff700 (LWP 520027)]
[New Thread 0x7fffdc9ff700 (LWP 520028)]
[Detaching after vfork from child process 520029]
[Thread 0x7fffe4fff700 (LWP 520014) exited]
[Detaching after vfork from child process 520030]
[Thread 0x7fffe37ff700 (LWP 520016) exited]
[Detaching after vfork from child process 520031]
[Thread 0x7fffe1dff700 (LWP 520018) exited]
[Detaching after vfork from child process 520032]
[Thread 0x7fffe15ff700 (LWP 520019) exited]
[Detaching after vfork from child process 520033]
[Thread 0x7fffe25ff700 (LWP 520017) exited]
[Detaching after vfork from child process 520034]
[Thread 0x7fffe0dff700 (LWP 520020) exited]
--Type <RET> for more, q to quit, c to continue without paging--
[Switching to Thread 0x7fffe3e0f700 (LWP 520015)]

Thread 7 "opt compiler_bu" hit Breakpoint 1, gcc::jit::recording::context::add_error (
    this=0x7fffe8886300, loc=0x7fffdf50d040, fmt=fmt@entry=0x7fffecbf822d "%s")
    at ../../../gcc/gcc/jit/jit-recording.c:1457
1457    {
(gdb) p loc->m_line
$1 = 2219
(gdb)
```

This is interesting. I tried the GCC build again without `--disable-bootstrap`, and the GCC build _itself_ failed (eventually) with this error:

```
../../../gcc-antoyo/gcc/jit/jit-playback.c: In member function ‘gcc::jit::playback::lvalue* gcc::jit::playback::context::new_global_with_value(gcc::jit::playback::location*, gcc_jit_global_kind, gcc::jit::playback::type*, gcc::jit::playback::rvalue*, const char*)’:
../../../gcc-antoyo/gcc/jit/jit-playback.c:715:8: error: unused variable ‘inner_type’ [-Werror=unused-variable]
  715 |   tree inner_type = type->as_tree ();
      |        ^~~~~~~~~~
```

I don't know what to make of this, but it suggests my GCC build is misconfigured or something. I have never built GCC before.

--- a/Readme.md
+++ b/Readme.md
@@ -111,6 +111,8 @@ Or add a breakpoint to `add_error` in gdb and print the line number using:
 p loc->m_line
 ```
 
+To get the `rustc` command to run in `gdb`, add the `--verbose` flag to `cargo build`.
+
 ### How to use a custom-build rustc
 
  * Build the stage2 compiler (`rustup toolchain link debug-current build/x86_64-unknown-linux-gnu/stage2`).

--- a/src/back/write.rs
+++ b/src/back/write.rs
@@ -1,4 +1,4 @@
-use std::fs;
+use std::{env, fs};
 
 use gccjit::OutputKind;
 use rustc_codegen_ssa::{CompiledModule, ModuleCodegen};
@@ -42,17 +42,17 @@ pub(crate) unsafe fn codegen(cgcx: &CodegenContext<GccCodegenBackend>, _diag_han
                 let _timer = cgcx
                     .prof
                     .generic_activity_with_arg("LLVM_module_codegen_emit_obj", &module.name[..]);
-                match &*module.name {
-                    "std_example.7rcbfp3g-cgu.15" => {
-                        println!("Dumping reproducer {}", module.name);
-                        let _ = fs::create_dir("/tmp/reproducers");
-                        // FIXME(antoyo): segfault in dump_reproducer_to_file() might be caused by
-                        // transmuting an rvalue to an lvalue.
-                        // Segfault is actually in gcc::jit::reproducer::get_identifier_as_lvalue
-                        context.dump_reproducer_to_file(&format!("/tmp/reproducers/{}.c", module.name));
-                        println!("Dumped reproducer {}", module.name);
-                    },
-                    _ => (),
+                if env::var("CG_GCCJIT_DUMP_MODULE_NAMES").as_deref() == Ok("1") {
+                    println!("Module {}", module.name);
+                }
+                if env::var("CG_GCCJIT_DUMP_MODULE").as_deref() == Ok(&module.name) {
+                    println!("Dumping reproducer {}", module.name);
+                    let _ = fs::create_dir("/tmp/reproducers");
+                    // FIXME(antoyo): segfault in dump_reproducer_to_file() might be caused by
+                    // transmuting an rvalue to an lvalue.
+                    // Segfault is actually in gcc::jit::reproducer::get_identifier_as_lvalue
+                    context.dump_reproducer_to_file(&format!("/tmp/reproducers/{}.c", module.name));
+                    println!("Dumped reproducer {}", module.name);
                 }
                 context.compile_to_file(OutputKind::ObjectFile, obj_out.to_str().expect("path to str"));
             }

--- a/src/intrinsic/mod.rs
+++ b/src/intrinsic/mod.rs
@@ -525,7 +525,7 @@ impl<'a, 'gcc, 'tcx> Builder<'a, 'gcc, 'tcx> {
 
         let value =
             if result_type.is_signed(self.cx) {
-                self.context.new_bitcast(None, value, typ)
+                self.context.new_cast(None, value, typ)
             }
             else {
                 value
@@ -689,7 +689,7 @@ impl<'a, 'gcc, 'tcx> Builder<'a, 'gcc, 'tcx> {
                 },
             };
 
-        self.context.new_bitcast(None, result, result_type)
+        self.context.new_cast(None, result, result_type)
     }
 
     fn count_leading_zeroes(&self, width: u64, arg: RValue<'gcc>) -> RValue<'gcc> {
@@ -740,6 +740,11 @@ impl<'a, 'gcc, 'tcx> Builder<'a, 'gcc, 'tcx> {
                 let not_low = self.context.new_unary_op(None, UnaryOp::LogicalNegate, self.u64_type, low);
                 let not_low_and_not_high = not_low & not_high;
                 let index = not_high + not_low_and_not_high;
+                // NOTE: the following cast is necessary to avoid a GIMPLE verification failure in
+                // gcc.
+                // TODO(antoyo): do the correct verification in libgccjit to avoid an error at the
+                // compilation stage.
+                let index = self.context.new_cast(None, index, self.i32_type);
 
                 let res = self.context.new_array_access(None, result, index);
 
@@ -763,7 +768,7 @@ impl<'a, 'gcc, 'tcx> Builder<'a, 'gcc, 'tcx> {
         let arg =
             if result_type.is_signed(self.cx) {
                 let new_type = result_type.to_unsigned(self.cx);
-                self.context.new_bitcast(None, arg, new_type)
+                self.context.new_cast(None, arg, new_type)
             }
             else {
                 arg
@@ -815,10 +820,15 @@ impl<'a, 'gcc, 'tcx> Builder<'a, 'gcc, 'tcx> {
                 let not_high = self.context.new_unary_op(None, UnaryOp::LogicalNegate, self.u64_type, high);
                 let not_low_and_not_high = not_low & not_high;
                 let index = not_low + not_low_and_not_high;
+                // NOTE: the following cast is necessary to avoid a GIMPLE verification failure in
+                // gcc.
+                // TODO(antoyo): do the correct verification in libgccjit to avoid an error at the
+                // compilation stage.
+                let index = self.context.new_cast(None, index, self.i32_type);
 
                 let res = self.context.new_array_access(None, result, index);
 
-                return self.context.new_bitcast(None, res, result_type);
+                return self.context.new_cast(None, res, result_type);
             }
             else {
                 unimplemented!("count_trailing_zeroes for {:?}", arg_type);
@@ -832,7 +842,7 @@ impl<'a, 'gcc, 'tcx> Builder<'a, 'gcc, 'tcx> {
                 arg
             };
         let res = self.context.new_call(None, count_trailing_zeroes, &[arg]);
-        self.context.new_bitcast(None, res, result_type)
+        self.context.new_cast(None, res, result_type)
     }
 
     fn int_width(&self, typ: Type<'gcc>) -> i64 {
@@ -846,7 +856,7 @@ impl<'a, 'gcc, 'tcx> Builder<'a, 'gcc, 'tcx> {
 
         let value =
             if result_type.is_signed(self.cx) {
-                self.context.new_bitcast(None, value, value_type)
+                self.context.new_cast(None, value, value_type)
             }
             else {
                 value
@@ -862,7 +872,7 @@ impl<'a, 'gcc, 'tcx> Builder<'a, 'gcc, 'tcx> {
             let low = self.context.new_cast(None, value, self.cx.ulonglong_type);
             let low = self.context.new_call(None, popcount, &[low]);
             let res = high + low;
-            return self.context.new_bitcast(None, res, result_type);
+            return self.context.new_cast(None, res, result_type);
         }
 
         // First step.
@@ -887,7 +897,7 @@ impl<'a, 'gcc, 'tcx> Builder<'a, 'gcc, 'tcx> {
         let value = left + right;
 
         if value_type.is_u8(&self.cx) {
-            return self.context.new_bitcast(None, value, result_type);
+            return self.context.new_cast(None, value, result_type);
         }
 
         // Fourth step.
@@ -898,7 +908,7 @@ impl<'a, 'gcc, 'tcx> Builder<'a, 'gcc, 'tcx> {
         let value = left + right;
 
         if value_type.is_u16(&self.cx) {
-            return self.context.new_bitcast(None, value, result_type);
+            return self.context.new_cast(None, value, result_type);
         }
 
         // Fifth step.
@@ -909,7 +919,7 @@ impl<'a, 'gcc, 'tcx> Builder<'a, 'gcc, 'tcx> {
         let value = left + right;
 
         if value_type.is_u32(&self.cx) {
-            return self.context.new_bitcast(None, value, result_type);
+            return self.context.new_cast(None, value, result_type);
         }
 
         // Sixth step.
@@ -919,7 +929,7 @@ impl<'a, 'gcc, 'tcx> Builder<'a, 'gcc, 'tcx> {
         let right = shifted & mask;
         let value = left + right;
 
-        self.context.new_bitcast(None, value, result_type)
+        self.context.new_cast(None, value, result_type)
     }
 
     // Algorithm from: https://blog.regehr.org/archives/1063

